pennylane/ops/op_math/exp.py

```diff
@@ -276,6 +328,116 @@ def pow(self, z):
         return Exp(self.base, self.coeff * z)
 
     def simplify(self):
-        if isinstance(self.base, qml.ops.op_math.SProd):  # pylint: disable=no-member
-            return Exp(self.base.base.simplify(), self.coeff * self.base.scalar)
-        return Exp(self.base.simplify(), self.coeff)
+        new_base = self.base.simplify()
+        if isinstance(new_base, qml.ops.op_math.SProd):  # pylint: disable=no-member
+            return Exp(new_base.base, self.coeff * new_base.scalar)
+        return Exp(new_base, self.coeff)
+
+    def generator(self):
+        r"""Generator of an operator that is in single-parameter-form.
+
+        For example, for operator
+
+        .. math::
+
+            U(\phi) = e^{i\phi (0.5 Y + Z\otimes X)}
+
+        we get the generator
+
+        >>> U.generator()
+          (0.5) [Y0]
+        + (1.0) [Z0 X1]
+
+        """
+        for op in [self, qml.simplify(self)]:
```

Comment:
I don't think we should be simplifying without telling the user this is happening, It could be very expensive. Could we have a kwarg to determine if we simplify first or not?

Reply:
Do we really need the kwarg? How about raising a warning to make the user aware? 🤔

```
if self.base.is_hermitian and not np.real(self.coeff):
    return self.base

op = qml.simplify(self)
if op.base.is_hermitian and not np.real(op.coeff):
    warnings.warn("Generator has been computed for the simplified form of Operation")
    return op.base

raise GeneratorUndefinedError(
    f"Exponential with coefficient {self.coeff} and base operator {self.base} does not have a generator."
)
```

Reply:
I don't think we should use a warning, because there is no way the user can avoid it (i.e there is no corrective behaviour for the warning except not calling the method). I think the kwarg makes sense, otherwise we can also not simplify, and catch any errors that are caused due to not simplifying, then raise an error telling users to try simplifying first.

Reply:
I guess the motivating example for this was if the coefficient is real and the base operator has an imaginary component. @Jaybsoni 's example was `base = qml.s_prod(1j, qml.Identity(wires=0))` and `coeff=1`. In that case, if you simplify, you get a correct generator, but if you just check `.is_hermitian` and `not np.real`, in the original formulation of the function you got a `GeneratorUndefinedError`. In this case, we wouldn't want it to proceed with `op.base` as a generator without first simplifying and moving `1j` into the coefficient. 

For the Exp class I've put it back to just raising an error if it doesn't seem to obviously fit the pattern the function is expecting, and suggested in the error message that simplifying the operator might help. I've added a generator function on Evolution that only raises and warning and then proceeds, since the idea with that class is we assume they intentionally defined the base operator to be a generator, unlike if you define an Exp.

Does that seem reasonable?

Reply:
Perfect thanks !